src/main/java/com/google/devtools/build/lib/sandbox/cgroups/VirtualCGroup.java

```diff
@@ -0,0 +1,208 @@
+import com.google.devtools.build.lib.events.Event;
+import com.google.devtools.build.lib.events.EventHandler;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.Controller;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v1.LegacyCpu;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v1.LegacyMemory;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v2.UnifiedCpu;
+import com.google.devtools.build.lib.sandbox.cgroups.controller.v2.UnifiedMemory;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Scanner;
+import java.util.concurrent.ConcurrentLinkedQueue;
+
+
+/**
+ * This class creates and exposes a virtual cgroup for the bazel process and allows creating
+ * child cgroups. Resources are exposed as {@link Controller}s, each representing a
+ * subsystem within the virtual cgroup and that could in theory belong to different real cgroups.
+ */
+@AutoValue
+public abstract class VirtualCGroup {
+    private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+    private static final File PROC_SELF_MOUNTS_PATH = new File("/proc/self/mounts");
+    private static final File PROC_SELF_CGROUP_PATH = new File("/proc/self/cgroup");
+
+    @Nullable
+    private static volatile VirtualCGroup instance;
+
+    @Nullable
+    public abstract Controller.Cpu cpu();
+    @Nullable
+    public abstract Controller.Memory memory();
+    public abstract ImmutableSet<Path> paths();
+
+    private final Queue<VirtualCGroup> children = new ConcurrentLinkedQueue<>();
+
+    public static VirtualCGroup getInstance() {
+        if (instance == null) {
+            synchronized (VirtualCGroup.class) {
+                if (instance == null) {
+                    try {
+                        instance = create().child("bazel_" + ProcessHandle.current().pid() + ".slice");
+                    } catch (IOException e) {
+                        logger.atInfo().withCause(e).log("Failed to create root cgroup");
+                        instance = NULL;
+                    }
+                    Runtime.getRuntime().addShutdownHook(new Thread(VirtualCGroup::deleteInstance));
+                }
+            }
+        }
+        return instance;
+    }
+
+    public static void deleteInstance() {
+        if (instance != null) {
+            synchronized (VirtualCGroup.class) {
+                if (instance != null) {
+                    instance.delete();
+                    instance = null;
+                }
+            }
+        }
+    }
+
+    static public VirtualCGroup NULL = new AutoValue_VirtualCGroup(null, null, ImmutableSet.of());
+
+    public static VirtualCGroup create() throws IOException {
```

Comment:
nit: rename this to something like `createRootCgroup` so that we can distinguish what its actually creating? Also, use package private access since this is the exposed API here should be `getInstance`.  (same for the methods below).

Reply:
Done